src/main/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacade.java

```diff
@@ -436,6 +436,72 @@ public final Response getQuizAttempt(@Context final HttpServletRequest httpServl
         }
     }
 
+    /**
+     * Mark a QuizAttempt as complete (or incomplete).
+     *
+     * Only owner and group managers for the quiz assignment can mark it incomplete.
+     * If it is a self-taken quiz, you cannot mark it incomplete, as that would make it easier to tweak individual
+     * questions to find the right answers.
+     *
+     * @param httpServletRequest
+     *            - so that we can extract user information.
+     * @param quizAttemptId
+     *            - the ID of the quiz the user wishes to attempt.
+     * @param markIncomplete
+     *            - optional boolean for group manager to mark incomplete.
+     * @return Confirmation or an error.
+     */
+    @POST
+    @Path("/attempt/{quizAttemptId}/complete")
+    @Produces(MediaType.APPLICATION_JSON)
+    @GZIP
+    @ApiOperation(value = "Mark a QuizAttempt as complete (or incomplete).")
+    public final Response completeQuizAttempt(@Context final HttpServletRequest httpServletRequest,
+                                              @PathParam("quizAttemptId") final Long quizAttemptId,
+                                              @FormParam("markIncomplete") final Boolean markIncomplete) {
+        try {
+            RegisteredUserDTO user = this.userManager.getCurrentRegisteredUser(httpServletRequest);
+
+            QuizAttemptDTO quizAttempt = getQuizAttempt(quizAttemptId);
+
+            QuizAssignmentDTO assignment = getQuizAssignment(quizAttempt);
+
+            boolean requestedMarkComplete = markIncomplete == null || markIncomplete == false;
+
+            Boolean markComplete;
+
+            if (assignment != null && canManageAssignment(assignment, user) && requestedMarkComplete == false) {
+                markComplete = false;
+            } else if (quizAttempt.getUserId().equals(user.getId()) && requestedMarkComplete == true) {
+                markComplete = true;
+            } else {
+                // Heuristic for most helpful error message.
+                return new SegueErrorResponse(Status.FORBIDDEN, isUserTeacherOrAbove(userManager, user)
+                    ? "You can only mark assignments incomplete for groups you own or manage."
+                    : "You cannot complete someone else's quiz.").toResponse();
+            }
+
+            if ((quizAttempt.getCompletedDate() != null) == markComplete) {
```

Comment:
👀 

(This is both very clever and also hard to read without thinking, maybe have a `currentCompletionStatus` variable or a comment?)

Reply:
This has gone away with the split into two endpoints.